core/test/solver/cgs.cpp

```diff
@@ -0,0 +1,193 @@
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <core/solver/cgs.hpp>
+
+
+#include <typeinfo>
+
+
+#include <gtest/gtest.h>
+
+
+#include <core/base/executor.hpp>
+#include <core/matrix/dense.hpp>
+
+
+namespace {
+
+
+class Cgs : public ::testing::Test {
+protected:
+    using Mtx = gko::matrix::Dense<>;
+    using Solver = gko::solver::Cgs<>;
+
+    Cgs()
+        : exec(gko::ReferenceExecutor::create()),
+          mtx(gko::initialize<Mtx>(
+              {{2, -1.0, 0.0}, {-1.0, 2, -1.0}, {0.0, -1.0, 2}}, exec)),
+          cgs_factory(gko::solver::CgsFactory<>::create(exec, 3, 1e-6)),
+          solver(cgs_factory->generate(mtx))
+    {}
+
+    std::shared_ptr<const gko::Executor> exec;
+    std::shared_ptr<Mtx> mtx;
+    std::unique_ptr<gko::solver::CgsFactory<>> cgs_factory;
+    std::unique_ptr<gko::LinOp> solver;
+
+    static void assert_same_matrices(const Mtx *m1, const Mtx *m2)
+    {
+        ASSERT_EQ(m1->get_num_rows(), m2->get_num_rows());
+        ASSERT_EQ(m1->get_num_cols(), m2->get_num_cols());
+        for (gko::size_type i = 0; i < m1->get_num_rows(); ++i) {
+            for (gko::size_type j = 0; j < m2->get_num_cols(); ++j) {
+                EXPECT_EQ(m1->at(i, j), m2->at(i, j));
+            }
+        }
+    }
+};
+
+
+TEST_F(Cgs, CgsFactoryKnowsItsExecutor)
+{
+    ASSERT_EQ(cgs_factory->get_executor(), exec);
+}
+
+
+TEST_F(Cgs, CgsFactoryKnowsItsIterationLimit)
+{
+    ASSERT_EQ(cgs_factory->get_max_iters(), 3);
+}
+
+
+TEST_F(Cgs, CgsFactoryKnowsItsRelResidualGoal)
+{
+    ASSERT_EQ(cgs_factory->get_rel_residual_goal(), 1e-6);
+}
+
+
+TEST_F(Cgs, CgsFactoryCreatesCorrectSolver)
+{
+    ASSERT_EQ(solver->get_num_rows(), 3);
+    ASSERT_EQ(solver->get_num_cols(), 3);
+    ASSERT_EQ(solver->get_num_stored_elements(), 9);
+    auto cgs_solver = static_cast<Solver *>(solver.get());
+    ASSERT_EQ(cgs_solver->get_max_iters(), 3);
+    ASSERT_EQ(cgs_solver->get_rel_residual_goal(), 1e-6);
+    ASSERT_NE(cgs_solver->get_system_matrix(), nullptr);
+    ASSERT_EQ(cgs_solver->get_system_matrix(), mtx);
+}
+
+
+TEST_F(Cgs, CanBeCopied)
+{
+    auto copy = cgs_factory->generate(Mtx::create(exec));
+
+    copy->copy_from(solver.get());
+
+    ASSERT_EQ(copy->get_num_rows(), 3);
+    ASSERT_EQ(copy->get_num_cols(), 3);
+    ASSERT_EQ(copy->get_num_stored_elements(), 9);
+    auto copy_mtx = static_cast<Solver *>(copy.get())->get_system_matrix();
+    assert_same_matrices(static_cast<const Mtx *>(copy_mtx.get()), mtx.get());
```

Comment:
minor: this comes from `CG`, but right now I think we can do this test using `ASSERT_MTX_NEAR` with a precision of 0.

Reply:
assert_same_matrices(static_cast<const Mtx *>(copy_mtx.get()), mtx.get());
->
ASSERT_MTX_NEAR(copy_mtx.get(), mtx.get(), 0.0);
?

Reply:
You actually should not need the whole `get` and `static_cast` I think. `ASSERT_MTX_NEAR` should handle the unique pointers properly.

Reply:
ASSERT_MTX_NEAR(copy_mtx, mtx, 0.0); ?

Reply:
My bad, you can try but it may not work. It really relies on dense conversion and you did not add that yet.

Reply:
so I guess we leave it like this for now? @gflegar, what do you think?

Reply:
Well isn't your matrix one of the types that has a dense conversion?

I don't see why we couldn't use `ASSERT_MTX_NEAR(static_cast<const Mtx*>(copy_mtx.get()), mtx, 0.0);` here...

Note that we do need a `static_cast` for `copy_mtx`. This is required because `ASSERT_MTX_NEAR` needs to know the `ValueType` of the matrix, and `LinOp` doesn't carry this information.

For the second one, @tcojean is right, the macro will take care of conversions and smart pointers (it also takes care of copying the matrix back to the CPU if it's not already there).

Reply:
Ah, actually, this is in `core/test`. For conversions we need one of the modules compiled (for now we need `reference`), so `ASSERT_MTX_NEAR` will result in a `NotCompiled` error here (we have to fix that at some point). So for now, let's just leave it as it is.